### ISSUE TITLE
trajectory slice operator doesn't return generator with a len function.

### ISSUE THREAD

**Issue by kain88-de** [MDAnalysis/mdanalysis#1894]
### Expected behaviour

`len(u.trajectory[::2])` returns half max trajectory length

### Actual behaviour

The generator doesn't have a len method.

This should in theory be possible. The `__getitem__` operator will have to return an object that has a `__len__` method. I'm not sure how much work this is though.

### Currently version of MDAnalysis:
(run `python -c "import MDAnalysis as mda; print(mda.__version__)"`)
0.18.0

**Comment by richardjgowers** [comment]
So there's currently 3 ways you could be iterating a trajectory (assuming no format specific Readers are doing their own thing....)

1) Standard `__iter__`

I think this one magically works because we've implemented `__len__` and `__getitem__`, so Python just calls `(Reader[i] for i in range(len(Reader)))`

https://github.com/MDAnalysis/mdanalysis/blob/8bab35993272ad9f1c8ad4893f85bd29928450a8/package/MDAnalysis/coordinates/base.py#L1285-L1288

2) Sliced iteration

What @kain88-de linked above, `Reader[x:y:z]`

https://github.com/MDAnalysis/mdanalysis/blob/8bab35993272ad9f1c8ad4893f85bd29928450a8/package/MDAnalysis/coordinates/base.py#L1365-L1382

3) Finally numpy style iteration

Can be either `Reader[[0, 5, 8, 10]]` or a boolean mask `Reader[[True, False, False, True]]`

https://github.com/MDAnalysis/mdanalysis/blob/8bab35993272ad9f1c8ad4893f85bd29928450a8/package/MDAnalysis/coordinates/base.py#L1329-L1333

**Comment by jbarnoud** [comment]
`__getitem__` handles case 2 and 3 from @richardjgowers's comment. For now we return a generator function, but we could return an iterator class without too much trouble. That class should have an `__iter__` method that would replace the current generator we return now, and the `__len__` method that @kain88-de calls for.

Out of curiosity, @kain88-de, what is your use case to need the length of a trajectory slice?

**Comment by richardjgowers** [comment]
Len(u.traj[10::20]) is pretty cool.

I suspect Max is thinking of parallel analysis and load sharing.

**Comment by kain88-de** [comment]
The len method is our use case. @lukas-stelzl  had this problem currently and he is using `len(np.arange(u.trajectory.n_frames)[:stop:step])` to get the same result. I can also see this useful in other parts to precalculate the expected size of result arrays.

**Comment by mimischi** [comment]
I also stumbled on this limitation. Usually I use `tqdm_notebook` to show a progress bar. Providing the `total` argument, one gets an estimate of how long the process will take to complete. So far I also solved it with the `np.arange` line mentioned above.

Would be great, if one would retrieve the actual number of frames after slicing. Although the work around is not too hard, I was actually expecting the `.n_frames` method to work, even on sliced trajectories.
